xonsh/platform.py

```diff
@@ -0,0 +1,141 @@
+""" Module for platform-specific constants and implementations """
+
+import os
+import platform
+import sys
+
+try:
+    import distro
+except ImportError:
+    distro = None
+
+
+#
+# OS
+#
+
+ON_DARWIN = platform.system() == 'Darwin'
+ON_LINUX = platform.system() == 'Linux'
+ON_WINDOWS = platform.system() == 'Windows'
+
+ON_POSIX = (os.name == 'posix')
+
+
+
+#
+# Python & packages
+#
+
+PYTHON_VERSION_INFO = sys.version_info[:3]
+
+if PYTHON_VERSION_INFO < (3, 5, 0):
+    from pathlib import Path
+
+    class DirEntry(Path):
+        def inode(self, *args, **kwargs):
+            raise NotImplementedError
+
+    def scandir(path):
+        """ Compatibility wrapper for os.scandir from Python 3.5+ """
+        return (DirEntry(x) for x in os.listdir(path))
+else:
+    from os import scandir
+
+
+ON_ANACONDA = any(s in sys.version for s in ('Anaconda', 'Continuum'))
+
+try:
+    import pygments
+except ImportError:
+    HAS_PYGMENTS, PYGMENTS_VERSION = False, None
+else:
+    HAS_PYGMENTS, PYGMENTS_VERSION = True, pygments.__version__
+
+
+try:
+    import readline
+except ImportError:
+    HAS_READLINE = False
+else:
+    HAS_READLINE = False
+
+
+try:
+    import prompt_toolkit
```

Comment:
The Python readline situation is really weird. [This note](https://docs.python.org/3/library/readline.html#module-readline) in the docs only starts to get at that.  But basically, on  Mac OSX readline is secretly replaced by libedit. On Windows, I am not sure it exists at all unless you pip install gnureadline.  Then I think sometimes the readline module from Python tries to spoof the input() behaviour with stdin, stdout, and stderr that you would get with readline manually.  

Basically, it is impossible to be sure that you actually have a true readline without importing it, which is why all of our `import readline` statements are in try-except blocks.

In any event, I think that it is important to only import the shell interface that will get used.  The exception to this rule is, as you noted, when `xonfig` is run and we are trying to figure out what a user actually has on their system.